docs/proposals/network-policy.md

```diff
@@ -0,0 +1,316 @@
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG. These are expected to augment this proposal in a backwards compatible way.
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta1 NetworkPolicy API object
+- A structure on the `Namespace` object to control policy, to be developed as an annotation for now.
+
+### Namespace changes
+
+The following objects will be defined on a Namespace Spec.
+>NOTE: In v1beta1 these objects will be implemented as an annotation.
+
+```go
+type IngressIsolationPolicy string
+
+const (
+	// Deny all ingress traffic to pods in this namespace.
+	DefaultDeny IngressIsolationPolicy = "DefaultDeny"
+) 
+
+// Standard NamespaceSpec object, modified to include a new
+// NamespaceNetworkPolicy field.
+type NamespaceSpec struct {
+	// This is a pointer so that it can be left undefined.
+	NetworkPolicy *NamespaceNetworkPolicy `json:"networkPolicy,omitempty"`
+}
+
+type NamespaceNetworkPolicy struct {
+	// Ingress configuration for this namespace.  This config is 
+	// applied to all pods within this namespace. For now, only 
+	// ingress is supported.  This field is optional - if not 
+	// defined, then the cluster default for ingress is applied.
+	Ingress *NamespaceIngressPolicy `json:"ingress,omitempty"`
+}
+
+// Configuration for ingress to pods within this namespace.
+// For now, this only supports specifying an isolation policy.
+type NamespaceIngressPolicy struct {
+	// The isolation policy to apply to pods in this namespace.
+	// Currently this field only supports "DefaultDeny", but could 
+	// be extended to support other policies in the future.  When set to DefaultDeny,
+	// pods in this namespace are denied ingress traffic by default.  When not defined,
+	// the cluster default ingress policy is applied (currently allow all). 
+	Isolation *IngressIsolationPolicy `json:"isolation,omitempty"` 
+}
+```
+
+```yaml
+kind: Namespace
+apiVersion: v1
+spec:
+  networkPolicy:
+    ingress:
+      isolation: DefaultDeny
+```
+
+The above structures will be represented in v1beta1 as a json encoded annotation like so:
+
+```yaml
+kind: Namespace
+apiVersion: v1
+metadata:
+  annotations:
+    net.beta.kubernetes.io/network-policy: "{\"ingress\": {\"isolation\": \"DefaultDeny\"}}" 
+```
+
+### NetworkPolicy Go Definition
+
+Once a Namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
```

Comment:
We should add a section covering why we did not choose:

a) a per-pod policy field (because of loose coupling, a la Services)
b) a per-service policy (this is still viable, IMO, and we should make sure reviewers have a chance to consider it)